drivers: can: Add driver support for TI MSPM0 G-Series MCAN module

santhosh-c-c opened the pull request:
This patch introduces driver support for the MCAN (CAN FD and classic CAN) module on TI's MSPM0 G-Series microcontrollers. The MCAN module supports both the classic CAN protocol and the newer CAN FD protocol, enabling flexible CAN communication.

Signed-off-by: Santhosh Charles santhosh@linumiz.com

santhosh-c-c committed:
dts: arm: ti: mspm0: g: add CAN-FD support for TI MSPM0 G-Series

Add CAN-FD devicetree nodes for MSPM0 G-Series SoCs that include
the peripheral: mspm0g310x, mspm0g350x and mspm0gx51x.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>
Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

santhosh-c-c committed:
drivers: clock_control: mspm0: add CANCLK source support

Add support for the CANCLK source for MSPM0. `mspm0_canclk_cfg`
config struct is used to get rate.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c committed:
dts: bindings: can: add DT bindings for TI MSPM0 G-Series MCAN module

Add devicetree bindings for the TI MSPM0 G-Series MCAN module.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>
Signed-off-by: Jackson Farley <j-farley@ti.com>

santhosh-c-c committed:
drivers: can: add driver support for TI MSPM0 G-Series MCAN module

Add driver support for MCAN module on TI's MSPM0 G-Series MCUs.
The MCAN module supports both classic CAN and CAN FD protocols.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>
Signed-off-by: Jackson Farley <j-farley@ti.com>
Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

santhosh-c-c commented:
> > Would you like review comments on the current state or should I wait until you move it out of draft?
> 
> @santhosh-c-c The PR description still says this is a draft PR. Is it a draft or would you like reviews?

@henrikbrixandersen The PR is ready for Review.

str4t0m commented on dts/bindings/can/ti,mspm0-canfd.yaml:
```diff
@@ -0,0 +1,38 @@
+      Selects the source clock for the CANCLK (MCAN_FCLK).
+      0: HFCLK
+      1: PLLCLK1
+      If not specified, defaults to PLLCLK1 (1).
```
TODO check naming details.

str4t0m commented on drivers/can/Kconfig.mspm0:
```diff
@@ -0,0 +1,31 @@
+	default y
+	depends on DT_HAS_TI_MSPM0_CANFD_ENABLED
+	select CAN_MCAN
+	select USE_MSPM0_DL_MCAN
```
Additionally `select PINCTRL`

str4t0m commented on drivers/can/Kconfig.mspm0:
```diff
@@ -0,0 +1,31 @@
+
+if CAN_MSPM0_CANFD
+
+config CAN_MAX_STD_ID_FILTER
```
TODO, I guees 128 STD and 64 EXT are not possible at the same time, should a check be added.

str4t0m commented on drivers/can/Kconfig.mspm0:
```diff
@@ -0,0 +1,31 @@
+	  Defines the maximum number of filters with extended ID (29-bit)
+	  that can be added by the application.
+
+endif # CAN_MSPM0G3XXX_CANFD
```
NIT: `CAN_MSPM0_CANFD` not `CAN_MSPM0G3XXX_CANFD`

str4t0m commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,276 @@
+		index, &can_mspm0_canfd_cfg_##index, &can_mspm0_canfd_ops,			\
+		&can_mspm0_canfd_cbs_##index);
+
+#define can_mspm0_canfd_DATA_INST(index)							\
```
Capitalize these defines.

str4t0m commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,276 @@
+	do {
+		DL_MCAN_enableModuleClock((MCAN_Regs *)config->ti_canfd_base);
+		DL_MCAN_getRevisionId((MCAN_Regs *)config->ti_canfd_base, &revid_MCAN0);
+	} while ((uint32_t)revid_MCAN0.scheme == 0x00);
```
TODO consider using a timeout.

henrikbrixandersen replied:
Could you please improve the comment here a bit? It's clear from the function names that this is fetching the revision ID, question is why this is needed...

str4t0m commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,276 @@
+
+	/* Init power */
+	DL_MCAN_reset((MCAN_Regs *)config->ti_canfd_base);
+	DL_MCAN_enablePower((MCAN_Regs *)config->ti_canfd_base);
```
TODO check hal, does it anything else?

henrikbrixandersen commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,276 @@
+		*rate = clock_rate;
+	}
+
+	return ret;
```
Nit: Could be simplified to:

```suggestion
	return 0;
```

henrikbrixandersen commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,276 @@
+#define CAN_MSPM0_CLK_SRC_PLLCLK1 1
+
+struct can_mspm0_canfd_config {
+	uint32_t ti_canfd_base;
```
You could store this as `MCAN_Regs *` instead and save the typecasts when setting registers.

henrikbrixandersen commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,276 @@
+#define can_mspm0_canfd_INST(index)								\
+	can_mspm0_canfd_IRQ_CFG_FUNCTION(index) can_mspm0_canfd_CFG_INST(index)			\
+		can_mspm0_canfd_DATA_INST(index) can_mspm0_canfd_DEVICE_INST(index)
+
```
I have not seen this construct before (especially wondering what's with the lower_UPPER case?). Please follow the naming used in other drivers and just inline these in one macro.

henrikbrixandersen commented on drivers/can/Kconfig.mspm0:
```diff
@@ -0,0 +1,31 @@
+	range 0 64
+	help
+	  Defines the maximum number of filters with extended ID (29-bit)
+	  that can be added by the application.
```
Where are these used?

santhosh-c-c replied:
The configs CONFIG_CAN_MAX_STD_ID_FILTER and CONFIG_CAN_MAX_EXT_ID_FILTER are used here:

CAN_MCAN_CALLBACKS_DEFINE(can_mspm0_canfd_cbs_##index,    CAN_MCAN_DT_INST_MRAM_TX_BUFFER_ELEMENTS(index),       CONFIG_CAN_MAX_STD_ID_FILTER, CONFIG_CAN_MAX_EXT_ID_FILTER);

henrikbrixandersen commented on dts/arm/ti/mspm0/g/mspm0g310x.dtsi:
```diff
@@ -1,3 +1,29 @@
+			 * # 3 TX FIFO elements =         3x8  = 24B
+			 * # 3 TX Buffer elements =       3x72 = 216B
+			 * # Total RAM                         = 816B
+			 */
```
I see no need for this comment.

henrikbrixandersen commented on dts/arm/ti/mspm0/g/mspm0g350x.dtsi:
```diff
@@ -1,3 +1,29 @@
+			 * # 3 TX FIFO elements =         3x8  = 24B
+			 * # 3 TX Buffer elements =       3x72 = 216B
+			 * # Total RAM                         = 816B
+			 */
```
Same.

henrikbrixandersen commented on dts/arm/ti/mspm0/g/mspm0gx51x.dtsi:
```diff
@@ -60,5 +60,49 @@
+			 * # 3 TX FIFO elements =         3x8  = 24B
+			 * # 3 TX Buffer elements =       3x72 = 216B
+			 * # Total RAM                         = 816B
+			 */
```
Same.

henrikbrixandersen commented on dts/arm/ti/mspm0/g/mspm0gx51x.dtsi:
```diff
@@ -60,5 +60,49 @@
+			 * # 3 TX Buffer elements =       3x72 = 216B
+			 * # Total RAM                         = 816B
+			 */
+			sample-point = <875>;
```
Same.

henrikbrixandersen commented on dts/arm/ti/mspm0/g/mspm0gx51x.dtsi:
```diff
@@ -60,5 +60,49 @@
+			 * # 3 TX FIFO elements =         3x8  = 24B
+			 * # 3 TX Buffer elements =       3x72 = 216B
+			 * # Total RAM                         = 816B
+			 */
```
Same.

henrikbrixandersen reviewed (changes requested):
Thanks! A few comments below.
Is there a board, where this can be enabled?

santhosh-c-c commented:
> Thanks! A few comments below. Is there a board, where this can be enabled?

This is for MSPM0 G series (310x, 350x, x51x). It includes only the CAN controller hardware, to test or use CAN communication you must connect an external CAN transceiver module.

santhosh-c-c commented on dts/arm/ti/mspm0/g/mspm0gx51x.dtsi:
```diff
@@ -60,5 +60,49 @@
+
+		canfd1: can@40510000 {
+			compatible = "ti,mspm0-canfd";
+			reg =  <0x40510000 0x8000>, <0x40510000 0x1000>, <0x40510000 0x400>;
```
I have just cross-checked the address with the official TRM for the x51x series, revision SLASFA2B – November 2024 (revised October 2025). According to this document, the address for MCAN1 remains 0x40510000. Could you please share your sources and the revision date you are referring to? This will help to confirm if there are differences due to newer or older revisions.

lsd-he360 replied:
I am referring to the same revision. By `m_can`, I mean the second entry of the `reg` array. If you are correct, then can you explain why the `m_can` address of `canfd0` is not `0x40508000` instead of `0x4050F000`?

str4t0m commented on dts/bindings/can/ti,mspm0-canfd.yaml:
```diff
@@ -0,0 +1,38 @@
+      1: PLLCLK1
+      If not specified, defaults to PLLCLK1 (1).
+
+  ti,divider:
```
I think as the clock divider really defines the divider ratio, and not some weird bit values like for example "0 divides by 1" etc going with the suggestion above should be fine.

str4t0m commented on drivers/can/Kconfig.mspm0:
```diff
@@ -0,0 +1,11 @@
+	default y
+	depends on DT_HAS_TI_MSPM0_CANFD_ENABLED
+	select CAN_MCAN
+	select USE_MSPM0_DL_MCAN
```
It should also `select PINCTRL`

str4t0m reviewed (changes requested):
Minor change request, and while you are at it you could reorder the clock control commit to come before the CAN driver commit.

santhosh-c-c committed:
dts: arm: ti: mspm0: add canclk node

Add the canclk node to the MSPM0 base dtsi to support CAN
clock configurations.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c commented on drivers/clock_control/clock_control_mspm0.c:
```diff
@@ -65,6 +65,14 @@ static struct mspm0_clk_cfg mspm0_lfclk_cfg = {
 };
 
+static struct mspm0_clk_cfg mspm0_canclk_cfg = {
+#if MSPM0_HFCLK_ENABLED && (DT_PROP(canfd0, ti_canclk_source) == 0)
```
Addressing this, I've added a canclk node in the devicetree. Its frequency will now be used to return the correct clock rate for both CANFD interfaces, ensuring support is not limited to canfd0.

henrikbrixandersen reviewed (changes requested):
Looks good, only two minor nits:

lsd-he360 commented:
I just want to mention that I had to override the `syspll` node defined in mspm0g.dtsi with the `clk1-div` set to `1` in order to get this CAN driver to work. What I did seems to match the steps outlined in the MSPM0 G-series TRM.

<img width="1030" height="1134" alt="image" src="https://github.com/user-attachments/assets/1239bcbb-c8a7-4ef0-9be5-569c6a7bcc37" />

henrikbrixandersen commented:
@santhosh-c-c Please avoid rebasing your branch on `main` unless absolutely necessary. It makes it very difficult to see the actual changes to this PR.

Is it ready for review? Do the test suites pass?

henrikbrixandersen commented on drivers/can/can_mspm0_canfd.c:
```diff
@@ -0,0 +1,312 @@
+	if (!WAIT_FOR(DL_MCAN_isMemInitDone(config->ti_canfd_base),
+		      MSPM0_MCAN_MEMINIT_TIMEOUT_US, k_busy_wait(1))) {
+		LOG_ERR("MSPM0 MCAN memory init failed");
+		return -ETIMEDOUT;
```
`init()` functions should typically return `-ENODEV` if the device initialization fails, but not a blocker from my side.

henrikbrixandersen commented on tests/drivers/can/api/boards/lp_mspm0g3507.overlay:
> The intent here is to keep CAN-FD disabled by default and let users opt-in via overlays. Moving this to the board would `enable` CAN-FD unconditionally for all users of `lp_mspm0g3507` which is not desirable.

That's not how Zephyr works. You enable the board peripherals in the board configuration, but drivers for e.g. CAN will not be enabled and built unless the application/sample/test/user set `CONFIG_CAN=y`.

> This test overlay exists purely to check CI and verify driver test suites. The board DTS already has the `canfd` node defined with the necessary params that is sufficient for users who want to enable it.

Having these overlays will not cause the driver nor the sample not anything to be built in CI or by twister unless explicitly requested. Twister, and thus CI, depends on the board listing a given hardware function to be supported in its YAML file, in this case the key is `can`. If that is not set, which it isn't here, the overlays have no value.

> Shall I drop the commits for the test suites?

They need to go to the board level.

santhosh-c-c committed:
boards: ti: lp_mspm0g3507: enable CAN-FD support

Enable CAN-FD peripheral on lp_mspm0g3507. Configure
canclk to use syspll as clock source and add pinctrl
for CAN TX/RX pins.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c committed:
boards: ti: lp_mspm0g3519: enable CAN-FD support

Enable CAN-FD peripheral on lp_mspm0g3519. Configure
canclk to use syspll as clock source and add pinctrl
for CAN TX/RX pins.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c commented:
> Why? As a user, I would much prefer being in control of the divisor for setting the time quanta. This is how all other CAN controllers with configurable prescalers work. What prompted you to change it?

I agree with your point. The intention was to simplify the configuration by having the driver handle the math, allowing the user to simply specify the target frequency. So moved the logic into the driver.

henrikbrixandersen reviewed (changes requested):
Looking good. Only a couple of nits left, see below.

Can you confirm that both `tests/drivers/can/api` and `tests/drivers/can/timing` test suites pass on these boards?

str4t0m reviewed (changes requested):
I think you need to rebase on main to pick up the sdk1.0 changes for the twister tests to pass.

henrikbrixandersen reviewed (approved):
Thanks! Please rebase on `main` to resolve the CI issue.

henrikbrixandersen commented:
Retriggered CI.
